bee-message/src/unlock/reference.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{error::ValidationError, unlock::UNLOCK_BLOCK_INDEX_RANGE, MessageUnpackError};
+
+use bee_packable::{PackError, Packable, Packer, UnpackError, Unpacker};
+
+use core::{
+    convert::{Infallible, TryFrom},
+    fmt,
+};
+
+/// Error encountered unpacking a [`ReferenceUnlock`].
+#[derive(Debug)]
+pub enum ReferenceUnlockUnpackError {
+    /// Validating a [`ReferenceUnlock`] failed.
+    ValidationError(ValidationError),
+}
+
+impl_wrapped_variant!(
+    ReferenceUnlockUnpackError,
+    ValidationError,
+    ReferenceUnlockUnpackError::ValidationError
+);
+
+impl fmt::Display for ReferenceUnlockUnpackError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::ValidationError(e) => write!(f, "{}", e),
+        }
+    }
+}
+
+/// An [`UnlockBlock`](crate::unlock::UnlockBlock) that refers to another [`UnlockBlock`](crate::unlock::UnlockBlock).
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+#[cfg_attr(feature = "serde1", derive(serde::Serialize, serde::Deserialize))]
+pub struct ReferenceUnlock(u16);
+
+impl ReferenceUnlock {
+    /// The unlock kind of a [`ReferenceUnlock`].
+    pub const KIND: u8 = 1;
+
+    /// Creates a new [`ReferenceUnlock`].
+    pub fn new(index: u16) -> Result<Self, ValidationError> {
+        if !UNLOCK_BLOCK_INDEX_RANGE.contains(&index) {
+            return Err(ValidationError::InvalidReferenceIndex(index));
+        }
+
+        Ok(Self(index))
+    }
+
+    /// Return the index of a [`ReferenceUnlock`].
```

Comment:
"Returns" (with an 's' at the end). Is this some kind of attention test? :troll:

Reply:
An attention test that I failed :trollface: